[JIT] Incremental updates to Alias Db in Mutation Remover pass

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35421 [JIT] Incremental updates to Alias Db in Mutation Remover pass**
* #33297 [JIT] Allow mutated values as functional graph inputs

This PR makes it so that we don't have to rebuild the entire alias db each time we remove a node in alias analysis.

Differential Revision: [D20922470](https://our.internmc.facebook.com/intern/diff/D20922470)

Commit:
[JIT] Incremental updates to Alias Db in Mutation Remover pass

eellison commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -309,13 +311,21 @@ struct MutationRemover {
-      // :(  TODO: incremental update ?
-      aliasDb_ = torch::make_unique<AliasDb>(graph_);
+      // TODO: don't strictly need to reset write cache, evaluate on models
+      aliasDb_->isWriteCacheStale_ = true;
```
I think it might actually be safe to leave `isWriteCacheStale_` to false here.

eellison commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -353,10 +363,33 @@ struct MutationRemover {
-      aliasDb_ = torch::make_unique<AliasDb>(graph_);
+      // now that we have removed a mutating op, the write cache is stale
+      // TODO: don't strictly need to reset write cache, evaluate on models
+      aliasDb_->isWriteCacheStale_ = true;
```
i think we can also leave `isWriteCacheStale_` false here.

- the mutated value will have one more write than it actually does, however it is safe to consider unmutated values as mutated.
- for the purposes of this pass, we never directly query if a value has writers. We just attempt to move two nodes together, which isn't affected by a slightly stale write cache. It is only affected by the uses / writes of a value between two nodes, and the written node is removed from the graph.

eellison replied:
What does `target invalidation` mean ?

A pass that mutates the graph already needs to be aware of how it is invalidating the aliasDb. I don't think it makes sense for the pass to generically maintain aliasDb state outside of its own invariants.

eellison replied:
I might do it as a follow up but leave it like this for now

suo reviewed (changes requested):
This PR introduces a lot of coupling between the mutation removal pass and AliasDb. We may want to think about larger restructurings that recognize that fact, for example it may make sense to move a bunch of this stuff into AliasDb.

eellison commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -119,6 +120,8 @@ class AliasDb {
   static bool mutableType(const Value* v);
   static bool mutableType(const TypePtr& type);
 
+  friend struct MutationRemover;
```
The options as far as I can tell are:
- not have incremental mutation do in-place updates. This means that the pass is O(n^2) (maybe even O(n^4)) and can not be on by default. 
- move all of this to AliasDb and make this publicly available. This is subtle code for very particular invariants, and I'm worried about exposing it generally. 
- current behavior. Make MutationRemover a friend of AliasDb. I think this is the least bad option. AliasDb & MutationRemover are maintained by the same people, and no one but us edits these files.
- Have MutationRemover have a private AliasDb class that inherits from the main AliasDb which exposes this functionality. I'm not sure exactly about this option, but we could then extend it to the `getAttr` behavior in freezing.

eellison reviewed (commented):
responded to your comment re: coupling

suo reviewed (approved):
Discussed offline—accepting this as we should unblock this work, but there are strong signs that the `AliasDb` abstraction is not sufficient—it was made under the assumption that the user would only want to make graph mutations that don't alter aliasing properties and does not expose the right handles to mutate its internal state.

I don't want to block this change on what I imagine will be a fairly involved refactor, but let's leave a TODO in the code and start mapping out a better design as part of the general functionalization work.

Commit:
Update on "[JIT] Incremental updates to Alias Db in Mutation Remover pass"


This PR makes it so that we don't have to rebuild the entire alias db each time we remove a node in alias analysis. 

[ghstack-poisoned]

Commit:
Update on "[JIT] Incremental updates to Alias Db in Mutation Remover pass"


This PR makes it so that we don't have to rebuild the entire alias db each time we remove a node in alias analysis.

Differential Revision: [D20922470](https://our.internmc.facebook.com/intern/diff/D20922470)

[ghstack-poisoned]
